### ISSUE TITLE
StreamBridge does not use/read internal kafka partition-count when sending to dynamic destinations

### ISSUE THREAD

**Issue by oliverfuehrer** [spring-cloud/spring-cloud-stream#2245]
Our application publishes events to multiple dynamic destinations (more then 100) with different partition-counts on several stages. We use spring-cloud-stream partitioning by setting "partition-key-expression".

We recognized that the partion-count for our producer destinations is only read from the configuration at startup time. Since we mostly use dynamic destinations, they are initialized with a default partition-count. In our previous implementation with BinderAwareChannelResolver the partition-count was updated at runtime with the real partition-count from the broker. But the new StreamBridge stack sticks with the initial/configured partition-count, and does not listen to the updates from the kafka-implementation.

Our workaround is to manually configure all partition-counts for all stages and topics manually (including adding the necessary bindings). This is a lot of configuration effort, and also requires application restarts when adding topics with preconfigured partition-counts. Another workaround was to use kafka-partitioning with message-key instead of scs-partitioning, but this uses too many resources on the broker.

We would like the new StreamBridge-stack to listen to partition-count updates from the kafka-implementation/broker like the old BinderAwareChannelResolver-stack did. This would greatly reduce configuration effort on our side, and get the flexibility back to add new producer destinations during runtime.

**Comment by olegz** [comment]
So, `BinderAwareChannelResolver` has no awareness of partition information or Kafka. It relies on `NewDestinationBindingCallback` which is also used by `StreamBridge`. 

The only difference is that once the destination has been used it is cached, so any updates that would be otherwise addressed by such callback will not manifest themselves. 

So I just want to confirm that this is the case you're talking about. Please confirm or provide more details

**Comment by oliverfuehrer** [comment]
We should distinguish beetween 2 different problems:
a) Using kafka broker topic partion count (not configured one) for partition calculation after startup (or first send call) 
b) Updating kafka broker topic partion count during runtime (NewDestinationBindingCallback?)

We would be very happy if "a)" would work ;)

Hint: If we use Kafka Message-Key "org.apache.kafka.clients.producer.internals.DefaultPartitioner" will be activated and partion will be computed correctly by using partion count from broker.

**Comment by olegz** [comment]
You're not answering my question. You've stated that it worked before with `BinderAwareChannelResolver`, so I am trying to determine the difference and the only thing I see is what I described before. 
OR.  .
Are you asking for additional feature/enhancement?

**Comment by oliverfuehrer** [comment]
Ok understand. The information that it worked with BinderAwareChannelResolver is from my colleage and we also have to discuss this (we will discuss it this afternoon). 

We would request for feature/enhancement that at least the initial partion count is read from kafka broker during "first send".

**Comment by olegz** [comment]
Cool, thanks. And just to add a little more context, if we remove the caching (I mentioned before) or make it configurable, or do something else clever to force callback on each call, then that may alleviate your issue

**Comment by oliverfuehrer** [comment]
Thank you. If you are also interested in our use case (or who we are) in general I would offer you to provide a web-ex session where you could get a short overview from us. If this is the case provide me with your email (for webex) and possible time slots. 🙂

**Comment by olegz** [comment]
That would work as well. Feel free to send me an invite at `ozhurakousky ar vmware`

**Comment by olegz** [comment]
So, the issue I believe was with the fact that the annotation-based programming model was relying on the process that involved `MessageConverterConfigurer` which added `PartitionInterceptor` to the binding channel. 
This was missed from `StreamBridge` and functional programing model, so this fix addresses it.

**Comment by oliverfuehrer** [comment]
Thank you for your fast help! I would expect that (with fixed version) kafka partion-count will be used when no partion count was configured for a specific destination - is this correct? What‘s about a possible configured default partition count?

**Comment by olegz** [comment]
@oliverfuehrer 
Here are the details. . .

When the binding starts, `KafkaTopicProvisioner` triggers producer destination provisioning, which also takes into consideration partitioning information. It does so by checking the following 3 things:
1. Min partition settings on the binder (`spring.cloud.stream.kafka.binder.minPartitionCount`) - default is 1.
2.  Partition count setting on the producer binding (`spring.cloud.stream.bindings.<binding-name>.producer.partitionCount`) - default is 1.
3. Provisioner also checks the number of actual partitions on the topic on the broker.

We find the effective partitions by finding the max of minPartition and partitionCount (1 & 2 above).
Then the binder, using the Kafka admin API query for retrieving the actual partition for the topic from the broker (3 above).
If 3 (whats on the server) is less then the effective partitions calculated, we increase the topic partitions on the broker.

The bottom line is that we do indeed read your configuration if present and yes it should take partition count from the broker if not configured via property.

Hope this helps, but as discussed before, please validate and let us now if something is not behaving the way you expect.

**Comment by oliverfuehrertsystems** [comment]
We just tested the latest 2021.0.0-SNAPSHOT in our local environment but the partition count from the kafka broker ist still not used.

We repeated the test with a much older software release and the correct partion count from kafka was used.

The old code used 'KafkaMessageChannelBinder.createProducerMessageHandle' during startup. 
This code updates 'DefaultPartitioningInterceptor' with the correct PartitionCount from Kafka (Line 424).

The StreamBridge from latest SNAPSHOT does not to use 'KafkaMessageChannelBinder', but creates the Binding and the 'DefaultPartitioningInterceptor' in a different way (see 'org.springframework.cloud.stream.function.StreamBridge.resolveDestination'). The 'DefaultPartitioningInterceptor' is not updated with the correct partionCount from kafka in this case.

**Comment by olegz** [comment]
@oliverfuehrer can you please test it again as I believe the above fix from @sobychacko addresses it.

**Comment by oliverfuehrertsystems** [comment]
We will test it tomorrow - at the moment I cannot see the changes in SNAPSHOT-2021.0.0 within our maven repository.

**Comment by oliverfuehrertsystems** [comment]
Is the Pull-Request still open?: https://github.com/spring-cloud/spring-cloud-stream/pulls

**Comment by oliverfuehrertsystems** [comment]
We are currently not able to test, because the changes have to be deployed in some public maven artifactory we can use from our build. 

We are currently using https://repo.spring.io/snapshot were it's not available (see https://repo.spring.io/ui/native/libs-snapshot/org/springframework/cloud/spring-cloud-stream/3.2.0-SNAPSHOT)

**Comment by olegz** [comment]
I see there is a build from 25th November. That should be good

**Comment by oliverfuehrertsystems** [comment]
> I see there is a build from 25th November. That should be good

Sorry, but you closed 3 days ago - means 26th ;)

**Comment by olegz** [comment]
I probably just closed the issue later. The 25th is correct build

**Comment by olegz** [comment]
Also, i just ran a new build so there are new snapshots from 29th

**Comment by olegz** [comment]
Guys, the 3.2 release is imminent (tomorrow or Wednesday) so, please confirm that it works for you as soon as you can if possible

**Comment by olegz** [comment]
Perfect, thank you!!

**Comment by roll57** [comment]
Hello, item is tag as Backport 3.1.x but last release 3.1.6 from december did not contains the fix.

Could you advice if the item will be backported ? Using Spring boot 2.4.x and Spring Cloud 2020.0.x I cannot use spring cloud stream 3.2.x

**Comment by olegz** [comment]
The commit is there 4224edb9
```
64fb0c82 (HEAD -> 3.1.x, upstream/3.1.x) Bumping versions to 3.1.7-SNAPSHOT after release
4da33e44 Going back to snapshots
8a56a4f1 (tag: v3.1.6) Update SNAPSHOT to 3.1.6
. . .
4224edb9 GH-2245 Ensure DefaultPartitionInterceptor is added to partitioned destinations managed by StreamBridge
a88973e1 GH-2206 Add resetLastError method to TestChannelBinder
. . .
```

**Comment by roll57** [comment]
Thank you for your answer, but I also would have expect the fix here:
https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/blob/bed956747040847553b49e5adbef7608a50db858/spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaMessageChannelBinder.java#L418

I dont see this commit being backported:
https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/commit/c9687189b7eaacc3d550b43185b3167c7b75a237

**Comment by roll57** [comment]
@olegz , do you think the fix will be backported on spring-cloud/spring-cloud-stream-binder-kafka 3.1.x ? See my previous remark.
Thanks

**Comment by olegz** [comment]
@sobychacko ^^

**Comment by sobychacko** [comment]
@roll57 We indeed miss backporting that change to the `3.1.x` branch of the Kafka binder. Apologies for missing it.  Now it is [backported](https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/commit/2fa0be02a1f1bb8f1b095a957c893107162ba89c). @olegz can advise when the next release of `3.1.x` is planned.

**Comment by zhanp0304** [comment]
Hi, spring-cloud-stream.version 3.1.6 is still not fixed the bug **'DefaultPartitioningInteceptor must be explicitly checked in order to avoid a CCE'.** 

background: I used the spring-cloud-dependencies version is 2020.0.5, and its spring-cloud-stream.version is defined as 3.1.6.

Can also fix this CCE in version 3.1.6? Please, thank you!

@sobychacko

**Comment by kubav182** [comment]
spring cloud version 2021.0.5 still not working, it sends everything into one partition

**Comment by sobychacko** [comment]
@kubav182 The `2021.0.5` release contains the Spring Cloud Stream version [3.2.6](https://github.com/spring-cloud/spring-cloud-release/blob/v2021.0.5/spring-cloud-dependencies/pom.xml#L32C32-L32C37), where this change [exists](https://github.com/spring-cloud/spring-cloud-stream/blob/v3.2.6/binders/kafka-binder/spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaMessageChannelBinder.java#L432). Could you verify your Spring Cloud Stream and Kafka binder versions? 
Since, both `3.1.x` and `4.0.x` are not [out of OSS support](https://spring.io/projects/spring-cloud-stream#support), any chance you can upgrade to the latest OSS supported GA version? That is `4.1.x` available via spring-cloud [2023.0.0](https://github.com/spring-cloud/spring-cloud-release/blob/v2023.0.0/spring-cloud-dependencies/pom.xml#L31)? 
If the problems still occur, could you open a new issue and reference this one? We will be happy to take a look and triage.

**Comment by kubav182** [comment]
@sobychacko Thanks for reply, upgrade to SB 3.2 and SC 2023.0.0 is in our plan, we wanted to replace deprecated classes (like Output annotation) first and use new functional style for consuming and StreamBridge for producing to make the upgrade easier. After this change we found all messages are in one partition. I checked versions and they are same as you mentioned. There was no change in version, so in this version it worked with "old style" but didn't work with "new style". 

As you mentioned there is no support for this version, so I will let you know after upgrade. For now we are ok with config property solution to set number of partitions.
